File: md_html_pdf.py
#!/usr/bin/env python3

import os
import sys

from markdown2 import markdown
from weasyprint import HTML, CSS


def read_file(filename):
    """Read the content of a file."""
    with open(filename, "r", encoding="utf-8") as file:
        return file.read()


def md_to_html_to_pdf(md_file):
    """Write HTML/PDF content to a file."""

    basename = md_file.replace(".md", "")
    resume_html = basename + ".html"
    resume_pdf = basename + ".pdf"

    extras = ["cuddled-lists", "tables", "footnotes"]
    md_content = read_file(md_file)

    html_content = markdown(md_content, extras=extras)
    with open(resume_html, "w", encoding="utf-8") as file:
        file.write(html_content)
    print("\nResume saved to " + resume_html)
    html = HTML(string=html_content)

    css = [CSS(filename="pdf.css")]
    html.write_pdf(resume_pdf, stylesheets=css)
    print("\nResume saved to " + resume_pdf)

    return


def main():
    if len(sys.argv) != 2:
        print("Usage: python md_html_pdf.py <resume in markdown>")
        sys.exit(1)

    if not os.path.exists(sys.argv[1]):
        raise FileNotFoundError("The file " + sys.argv[1] + " does not exist")

    md_file = sys.argv[1]
    md_to_html_to_pdf(md_file)


if __name__ == "__main__":
    main()

File: md_html_pdf.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/russross/blackfriday/v2"
)

func readFile(filename string) (string, error) {
	content, err := ioutil.ReadFile(filename)
	if err != nil {
		return "", fmt.Errorf("error reading file %s: %w", filename, err)
	}
	return string(content), nil
}

func convertMarkdownToHTML(markdownContent string) string {
	html := blackfriday.Run([]byte(markdownContent))
	return string(html)
}

func writeHTMLFile(htmlContent string, outputFile string) error {
	err := ioutil.WriteFile(outputFile, []byte(htmlContent), 0644)
	if err != nil {
		return fmt.Errorf("error writing HTML file %s: %w", outputFile, err)
	}
	fmt.Println("\nResume saved to", outputFile)
	return nil

}
func convertHTMLToPDF(inputHTML string, outputPDF string) error {

	// Create a temporary HTML file to pass to wkhtmltopdf
	tempHTMLFile, err := ioutil.TempFile("", "temp-html-*.html")
	if err != nil {
		return fmt.Errorf("error creating temporary HTML file: %w", err)
	}
	defer os.Remove(tempHTMLFile.Name()) // Clean up the temp file

	if _, err := tempHTMLFile.Write([]byte(inputHTML)); err != nil {
		return fmt.Errorf("error writing to temporary file: %w", err)
	}
	if err := tempHTMLFile.Close(); err != nil {
		return fmt.Errorf("error closing temp file: %w", err)
	}

	cmd := exec.Command("wkhtmltopdf", tempHTMLFile.Name(), outputPDF)
	err = cmd.Run()

	if err != nil {
		return fmt.Errorf("error converting to PDF: %w", err)
	}

	fmt.Println("\nResume saved to", outputPDF)
	return nil
}

func main() {
	if len(os.Args) != 2 {
		fmt.Println("Usage: go run main.go <resume in markdown>")
		os.Exit(1)
	}

	inputMarkdownFile := os.Args[1]

	if _, err := os.Stat(inputMarkdownFile); os.IsNotExist(err) {
		log.Fatalf("The file %s does not exist", inputMarkdownFile)
	}

	resumeContent, err := readFile(inputMarkdownFile)
	if err != nil {
		log.Fatalf("Could not read file: %v", err)
	}
	baseName := strings.TrimSuffix(inputMarkdownFile, filepath.Ext(inputMarkdownFile))
	resumeHTML := baseName + ".html"
	resumePDF := baseName + ".pdf"

	htmlContent := convertMarkdownToHTML(resumeContent)

	if err := writeHTMLFile(htmlContent, resumeHTML); err != nil {
		log.Fatalf("Error writing HTML file: %v", err)
	}

	if err := convertHTMLToPDF(htmlContent, resumePDF); err != nil {
		log.Fatalf("Error converting to PDF: %v", err)
	}

}
